### DISCUSSION TITLE
Can we distribute getAllChanges output, instead of stateful synchronization, for a client-server model?

### DISCUSSION THREAD

**Discussion by Sjoerd82** [automerge/automerge#1521] [General]
The function `getAllChanges` seems to spit out a distributable binary set of changes. Is this so? And is this in the same "compressed" binary format that some other functions, like `save`, produce?

The reason I ask is that I already have a client-server sync and changes system in place, based on PostgreSQL+PostgREST, which works well, and this since recently (unfortunatly) also includes a central node.js server (SPAs can't be trusted...). I'm considering various approaches, such as the one below, my main question is if `getAllChanges` returns me the _compressed binary_ (of which the documentation speaks in the `save`-function). And can it be used to send the client's changes to the server? (or am I going to shoot myself in the foot somewhere along the ride?)

The stateful syncing functions (`generateSyncMessage`, `receiveSyncMessage`) are probably more efficient, as the presumably are able to only transmit deltas between peers. In the diagram below, clients continually need to download the whole document (but only if it has changed). But, as I am currently a bit hesitant to include Websockets/WebRTC into the mix, I'd prefer to start exploring in a bit more simple way.

![image](https://github.com/automerge/automerge/assets/15013786/6ef2feec-52f6-4788-b742-5fd411cd6d7c)

Quick recap of this diagram. A client makes changes (debounced or with a SAVE button) using `change`. So we now have one or more changes that need to be sent to the central server for merging (locally we remain up-to-date by doing `applyChanges`, we use indexeddb here).

On the central server there is a node.js instance collecting all the incoming changes and continuously merging* them and making a new document available for other clients (other clients regularly poll a table with a list of timestamps to know if there's something new to download, and could also receive a notification of a change via SSE (server-sent events), if their browser and network support that. Once an updated doc is available, they can download it, and the process starts all over again.

*) Finally, a quick question about merging.. If I understand correctly, I would create a new doc based on the current state, apply the changes I received, and then `merge` that new doc with the current state doc, right?

Thanks!

**Comment by bokolob** [comment]
Hi. I use a similar approach. All clients send their changes through a websocket connection to the server. Server carefully saves them to postgres as is (without any merges, because I don't need document on server side yet). Then server sends notification via queue to all clients. Clients download  changes (not full document) from server.

I don't use sync protocol (maybe I will adopt it in future), but it works quite well. The only drawback is downloading of all changes by client when it connects first time. Especially when documents are big.  On the other hand I store everything in indexdDb, so it happens only for very first connection.

So server plays a role of a persisting network link.
